HttpClient should not miss Set-Cookie response headers

justinvp opened the pull request:
Recent changes due to PR #3199 caused `HttpClient` on Windows to miss `Set-Cookie` response headers when one of the header values was large enough to cause `WinHttpQueryHeaders` to fail with `ERROR_INSUFFICIENT_BUFFER`. This commit addresses the regression.

Added a new test to validate this change.

Fixes #6737.

cc: @davidsh @stephentoub 

@kichalla, FYI.

davidsh commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -466,6 +466,41 @@ public void MaxRequestContentBufferSize_Set_ThrowsPlatformNotSupportedException(
+
+            using (var client = new HttpClient(handler))
+            {
+                HttpResponseMessage response = await client.GetAsync(url);
```
You can add a `using` for the `HttpResponseMessage response = ... `

davidsh commented:
Test Innerloop Ubuntu Debug Build and Test
http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/ubuntu_debug_prtest/3335/ (server went offline)

davidsh commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -466,6 +466,41 @@ public void MaxRequestContentBufferSize_Set_ThrowsPlatformNotSupportedException(
+        public async Task GetAsync_ResponseWithSetCookieHeaders_AllCookiesRead()
+        {
+            // Cookie names/values, with one of the values deliberately over 128 characters,
+            // which is the size of WinHttpHandler's initial response header buffer size, to
```
>which is the size of WinHttpHandler's initial response header buffer size

Is there any buffer size issues that CurlHandler has? Will this test also cover any interesting edge cases for CurlHandler?

justinvp replied:
> Is there any buffer size issues that CurlHandler has?

Unlikely. Curl just hands us a pointer to a native memory, that we read. We don't have to allocate anything like with WinHttpHandler.

davidsh replied:
Then maybe this test really belongs in the System.Net.Http.WinHttpHandler/tests/FunctionalTests/WinHttpHandlerTest.cs.

While we usually try to have most Http tests in System.Net.Http library so that we exercise both WinHttpHandler and CurlHandler, this test feels like it just belongs in WinHttpHandler library only.  In which case, you would adjust the test code to do a `new WinHttpHandler` etc.

stephentoub replied:
> Then maybe this test really belongs in the System.Net.Http.WinHttpHandler/tests/FunctionalTests/WinHttpHandlerTest.cs.

While there may not be an explicit limit in CurlHandler, I think there's value in having the test for it as well, to exercise a longer cookie, a pair of cookies, that there's no issue with the various characters used in the cookies, etc. Most importantly, I think there's value in having as many different tests as possible doing the same thing on both platforms.

justinvp replied:
> Then maybe this test really belongs in the System.Net.Http.WinHttpHandler/tests/FunctionalTests/WinHttpHandlerTest.cs.

I had the same initial thought, but then figured it might be good to have this coverage for both WinHttpHandler and CurlHandler.

I've added some additional tests (via MemberData) and kept it in System.Net.Http for now.

justinvp commented:
I've been doing some more testing and have found that `WinHttpQueryHeaders` doesn't always advance the index. With an initial buffer of 128 chars, a `Set-Cookie` response header value of length:
 - 0-127 chars: succeeds, index advances from 0 to 1.
 - 128-255 chars: fails due to insufficient buffer, index advances from 0 to 1.
 - 256+ chars: fails due to insufficient buffer, index stays at 0.

I'm not sure why header values of length 128-255 cause `WinHttpQueryHeaders` to advance the index, whereas 256+ length header values do not.

Hence, I've removed the `Debug.Assert(index == originalIndex + 1);` that I had added before setting `index` back to `originalIndex`, because this isn't always the case. Now all we do for `ERROR_INSUFFICIENT_BUFFER` is always set `index = originalIndex` before retrying.

I've added some more tests that cover exact header value lengths at these edges. For now, I've left the tests in System.Net.Http.

stephentoub commented on src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs:
```diff
@@ -465,6 +464,71 @@ public void MaxRequestContentBufferSize_Set_ThrowsPlatformNotSupportedException(
+            new object[] { GenerateCookie(name: "foo", repeat: 'a', overallHeaderValueLength: 254) },
+            new object[] { GenerateCookie(name: "foo", repeat: 'a', overallHeaderValueLength: 255) },
+            new object[] { GenerateCookie(name: "foo", repeat: 'a', overallHeaderValueLength: 256) },
+            new object[] { GenerateCookie(name: "foo", repeat: 'a', overallHeaderValueLength: 257) },
```
Does HTTP define a max cookie size?  Considering adding one for that if so.  I see that CookieContainer defines a default max size... maybe ones around that.

Also, it'd be good to include a comment about why these values were chosen, so that someone reading the code in a few years understands why they're there.

davidsh replied:
I'm not aware of any size limitations. Would have to research the RFC's.

justinvp replied:
Thanks, guys. I'll submit a separate PR that adds a comment and I'll think about additional sizes to test. I likely won't get to it until Sunday/Monday.

stephentoub commented:
LGTM

davidsh commented:
LGTM
